src/main/java/com/intuit/graphql/orchestrator/batch/GraphQLServiceBatchLoader.java

```diff
@@ -212,6 +235,69 @@ private CompletableFuture<List<DataFetcherResult<Object>>> load(List<DataFetchin
           return batchResult;
         });
   }
+  public static <T> Predicate<T> distinctByFieldName(Function<? super T, ?> keyExtractor) {
+    Set<Object> seen = ConcurrentHashMap.newKeySet();
+    return t -> seen.add(keyExtractor.apply(t));
+  }
+
+  private boolean canOptimizeQuery(SelectionSet selectionSet){
+    if(serviceMetadata.hasFieldResolverDirective()) return false;
+    if(selectionSet == null) return true;
+    for( Selection selection : selectionSet.getSelections()){
+      if(selection.getClass() != Field.class
+              || ((Field) selection).getDirectives().size()!=0
+              || !canOptimizeQuery(((Field) selection).getSelectionSet()))
+        return false;
+    }
+    return true;
+  }
+  private SelectionSet mergeFilteredSelection(Field node, HashMap<String, Set<Field>> selectionSetTree){
+    Field field;
+    SelectionSet.Builder selectionSetBuilder = SelectionSet.newSelectionSet();
+    SelectionSet.Builder childSelectionSetBuilder = SelectionSet.newSelectionSet();
+    if (ObjectUtils.isEmpty(selectionSetTree.get(node.getName()))) { // leaf node
+      field = new Field(node.getName(), node.getArguments());
+      selectionSetBuilder.selection(field);
+      SelectionSet set = selectionSetBuilder.build();
+      return set;
+    }
+    List<SelectionSet> childrenSets = new ArrayList<>();
+    for( Field f : selectionSetTree.get(node.getName())) {
+      childrenSets.add(mergeFilteredSelection(f, selectionSetTree));
+    }
+    childrenSets
+            .stream()
+            .map(s -> s.getSelections())
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList())
+            .stream()
+            .forEach(childSelectionSetBuilder::selection);
+    field = new Field(node.getName(), node.getArguments(), childSelectionSetBuilder.build());
+    selectionSetBuilder.selection(field);
+
+    return selectionSetBuilder.build() ;
+  }
+
+  private void createSelectionSetTree(SelectionSet selectionSet, HashMap<String, Set<Field>> selectionSetMap){
+    selectionSet.getSelections().stream().forEach(field -> {
+      Field f = (Field) field;
+      if(f.getSelectionSet() != null){ // leaf node
```

Comment:
might want to update the comment to `// check if leaf node.  Initially it thought `f.getSelectionSet() != null` is a leaf node :)